### ISSUE TITLE
Best practice for URL versioning, filtering API operations based on users and grouping API operations

### ISSUE THREAD

**Issue by c-pinkerton** [springfox/springfox#1333]
Hi,

Just like to say first off that the Springfox project is excellent (it's been starred), it was so pleasant to start using it and I really appreciate the work that's going into it :-)

I am building a REST API using Spring Boot (v1.3.3) + Springfox (v2.4.0), and I have had a look through the existing issues, but couldn't find anything which answered all my questions.

My REST API has a few requirements that I'm not sure how best to fulfill, ideally the API should

1. Be versioned with multiple concurrent versions running (currently working towards a URL based versioning scheme, i.e. /v1/api/operationX, /v2/api/operationX). I'm expecting to create a separate RestController for each version of an API operation. The question here is how to display the versions in Swagger UI.

2. Show/hide some API operations in Swagger UI based on the authenticated user i.e. we might have a 'standard' user who can see API operations X and Y, and an admin user who can see operations X, Y and Z

3. Have related API operations logically grouped (ideally in a hierarchical structure) within Swagger UI. For example we might have several API operations to modify configuration of entities, it would be nice to have a structure something like

* config
  * rules
    * simple rules
        * simple rule X
           * GET...
           * PUT...
           * POST...
           * DELETE...
        * simple rule Y
           * GET...
           * PUT...
           * POST...
           * DELETE...
        * ...
    * complex rules
        * complex rule X
           * GET...
           * PUT...
           * POST...
           * DELETE...
        * complex rule Y
           * GET...
           * PUT...
           * POST...
           * DELETE...
        * ...
  * users
    * GET...
    * PUT...
    * POST...
    * DELETE...
  * other stuff....

Thanks for any help you can provide.

**Comment by dilipkrish** [comment]
@c-pinkerton thank you for your kind words! 👍 

> The question here is how to display the versions in Swagger UI.

My recommendation would be to create separate dockets for separate versions. That way you can group version separately. In the dockets you'd use separate RequestHandler predicates and/or Path predicates to determine the grouping

> Show/hide some API operations in Swagger UI based on the authenticated user

This is not really supported as part of the core library. You'd have to write something on your own for this.

> Have related API operations logically grouped (ideally in a hierarchical structure) within Swagger UI. 

The recommended way to do this is to use tagging. Kinda like how google does categorization, or even editor.swagger.io for that matter.

**Comment by c-pinkerton** [comment]
Thanks for the quick response :-) Are you aware of any examples of tagging API operations to produce a hierarchical structure in Swagger UI?

**Comment by dilipkrish** [comment]
@c-pinkerton tagging will not produce a hierarchical structure. Try out http://editor.swagger.io/#/ to see how that works

**Comment by c-pinkerton** [comment]
@dilipkrish So is it correct that you can only have single level of nesting when grouping by tags? I don't see any examples at http://editor.swagger.io/#/ which show more than one level deep

**Comment by dilipkrish** [comment]
*YES*. Hierarchical is not possible unless u write some custom code to make it hierarchical.

**Comment by c-pinkerton** [comment]
@dilipkrish Alright, thanks for confirming that for me :+1:

**Comment by ionutmohan** [comment]
Hi @dilipkrish,

Can you please give me additional details on how to achieve this:

**Show/hide some API operations in Swagger UI based on the authenticated user**
> This  is not really supported as part of the core library. You'd have to write something on your own for this.

It would be great if you can provide some steps or library classes which can be extended to be able to do this.

Thanks in advance!

**Comment by dilipkrish** [comment]
So the specification is an open to read for everyone one, you'd have to modify the [Swagger-ui](https://github.com/swagger-api/swagger-ui) and customize it to read the specification and only show the ones that user has access to.

**Comment by ionutmohan** [comment]
Thanks @dilipkrish for your help on this.

I would go this approach only if there is no other solution for working with an as-is version of the library. And as I was browsing the springfox questions and issues etc from this project I ended up trying to implement a solution similar to the one explained by andrea-sdl [here ](https://github.com/springfox/springfox/issues/2123)

The idea is to have multiple API groups. Let's say some are public some are internal.
I tried to implement my own SwaggerResourceProvider and inside the get do the filtering based on the authenticated user. If the user has privileges to see the internal API I would add it to the list otherwise I am returning only the public groups. 

The current issue I have is that this get method is hit only once when I access the page for the first time. At this time I don't have any Authorization set. So I will display only the public groups. After I go and do 'Authorize' I was expecting to have another call to the provider get method and this time I could've checked the user privileges and return also the internal groups if the user is permitted to see them. I am stuck now since the provider's get is not called at all after I use the 'Authorize' button from swagger-ui.

I would really appreciate any tipps on how to overcome this. Can I somehow trigger this get call after 'Authorize'?

**Comment by dilipkrish** [comment]
@ionutmohan Its not really built for that kind of dynamism. The idea is that the specification is what it is, it will not change based on authorization credentials. 

Swagger-ui on the other hand, as-is is designed to show the UI for everything, and uses the authentication mechanisms to determine if an action can be executed or not at runtime.

That said, you'd have to either make swagger ui dynamic (hard) or only allow access to the swagger resources after you authenticate (easy).
